spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+            1. Let |resultIndex| be the numerical value of |indexPromise|.
+            1. Let |context| be {{WindowSharedStorage}}'s {{Window}}'s [=Window/browsing context=].
+            1. [=Assert=] that |context| is not null.
+            1. Let |origin| be |context|'s [=active document=]'s [=url/origin=].
+            1. [=Assert=] that |origin| is not [=opaque origin|opaque=].
+            1. Let |remainingBudget| be the result of running [=determine remaining budget=] with |origin|.
+            1. [=Assert=] that |remainingBudget| is not undefined.
+            1. Let |listSize| be |urlList|'s [=list/size=].
+            1. Let |pendingBits| be the logarithm base 2 of |listSize|.
+            1. If |pendingBits| is greather than |remainingBudget|, set |resultIndex| to [=default index=].
+            1. Set |fencedFrameConfig|'s [=pending shared storage budget debit=] to |pendingBits|.
+            1. Set |fencedFrameConfig|'s [=FencedFrameConfig/url=] to |urlList|[|resultIndex|].
+        1. [=Upon rejection=] of |indexPromise|, set |fencedFrameConfig|'s [=FencedFrameConfig/url=] to |urlList|[[=default index=]].
+    1. Return |configPromise|.
+  </div>
+
+  ## The {{WorkletSharedStorage}} interface ## {#worklet-shared-storage}
+
+  The {{WorkletSharedStorage}} interface is as follows.
+
+<xmp class='idl'>
+  [Exposed=(SharedStorageWorklet)]
+  interface WorkletSharedStorage : SharedStorage {
+    Promise<DOMString> get(DOMString key);
+    Promise<unsigned long> length();
+    Promise<double> remainingBudget();
+
+    async iterable<DOMString, DOMString>;
+  };
 </xmp>
+
+  ### Worklet Setter/Deleter Methods ### {#worklet-setter}
+
+  <div algorithm>
+    The <dfn method for="WorkletSharedStorage">set(|key|, |value|, |options|)</dfn> method steps are:
+
+    1. Let |promise| be a new [=promise=].
+    1. Let |context| be {{WorkletSharedStorage}}'s {{SharedStorageWorkletGlobalScope}}'s [=outside settings=]'s [=target browsing context=].
+    1. [=Assert=] that |context| is not null.
+    1. If |key|'s [=string/length=] exceeds the [=key/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |value|'s [=string/length=] exceeds the [=value/maximum length=], reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |options| is invalid, reject |promise| with a "{{DataError}}" {{DOMException}}.
+    1. If |options| is empty, create |options|.
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |result| be null.
+        1. If |options|["`ignoreIfPresent`"] is true:
+            1. Let |currentValue| be the result of running [=shared storage database/retrieve an entry from the database=] with |context| and |key|.
+            1. If |currentValue| is not null and |currentValue| is not undefined, resolve |promise| as undefined.
+            1. Otherwise, if |currentValue| is null, reject |promise|.
```

Comment:
Also, this algorithm seems duplicated with the one above.  Should this be calling into the other one or a common algorithm they both use?  Is there a reason for them to be duplicated?  Are they slightly different?